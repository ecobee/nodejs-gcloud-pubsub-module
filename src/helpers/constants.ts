export const PUB_SUB_DEFAULT_RETRY_CODES = [
	10, // 'ABORTED'
	1, // 'CANCELLED',
	4, // 'DEADLINE_EXCEEDED'
	13, // 'INTERNAL'
	8, // 'RESOURCE_EXHAUSTED'
	14, // 'UNAVAILABLE'
	2, // 'UNKNOWN'
	5, // NOT_FOUND'
	7, // PERMISSION_DENIED
]

export const PUB_SUB_DEFAULT_BACKOFF_SETTINGS = {
	initialRetryDelayMillis: 100,
	retryDelayMultiplier: 1.3,
	maxRetryDelayMillis: 60000,
	initialRpcTimeoutMillis: 12000,
	rpcTimeoutMultiplier: 1.0,
	maxRpcTimeoutMillis: 30000,
	totalTimeoutMillis: 600000,
}

export const MESSAGE = 'message'
export const ERROR = 'error'
export const CLOSE = 'close'
