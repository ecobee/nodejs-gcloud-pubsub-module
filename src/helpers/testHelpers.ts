import { PublishOptions } from '@google-cloud/pubsub/build/src/topic'
import { GoogleAuthOptions } from '../interfaces/gcloud-pub-sub.interface'
import { PUB_SUB_DEFAULT_RETRY_CODES, PUB_SUB_DEFAULT_BACKOFF_SETTINGS } from './constants'
import { SubscriberOptions } from '@google-cloud/pubsub/build/src/subscriber'

export const mockGoogleAuthOptions: GoogleAuthOptions = {
	projectId: 'entitlement',
}

export const mockPublishOptions: PublishOptions = {
	gaxOpts: {
		retry: {
			retryCodes: PUB_SUB_DEFAULT_RETRY_CODES,
			backoffSettings: PUB_SUB_DEFAULT_BACKOFF_SETTINGS,
		},
	},
}

export const mockSubscriberOptions: SubscriberOptions = {
	flowControl: {
		maxMessages: 5,
		allowExcessMessages: false,
	},
}
