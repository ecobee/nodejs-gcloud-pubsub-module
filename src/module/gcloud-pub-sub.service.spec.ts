import { Test, TestingModule } from '@nestjs/testing'
import { Topic, PubSub } from '@google-cloud/pubsub'
import { GcloudPubSubService } from './gcloud-pub-sub.service'
import { mockGoogleAuthOptions, mockPublishOptions } from '../helpers/testHelpers'

jest.mock('@google-cloud/pubsub')

describe('GcloudPubSubService', () => {
	let service: GcloudPubSubService

	beforeEach(async () => {
		const module: TestingModule = await Test.createTestingModule({
			providers: [
				{
					provide: GcloudPubSubService,
					useFactory: async (): Promise<GcloudPubSubService> =>
						new GcloudPubSubService(mockGoogleAuthOptions, mockPublishOptions),
				},
			],
		}).compile()

		service = module.get<GcloudPubSubService>(GcloudPubSubService)
	})

	it('should be defined', () => {
		expect(service).toBeDefined()
	})

	describe('gcloudPubSubLib', () => {
		it('initializes gcloudPubSubLib', () => {
			expect(service.gcloudPubSubLib).toBeDefined()
		})
	})

	describe('publishMessage', () => {
		it('Publishes a new message', async () => {
			const topic = 'Homer'
			const data = 'You Tried Your Best and You Failed Miserably. The Lesson Is Never Try'
			const gcloudPubSubLibMock = {
				topic: jest.fn().mockReturnThis(),
				publish: jest.fn(buffer => {
					expect(buffer).toMatchSnapshot()
				}),
			}
			service.gcloudPubSubLib = gcloudPubSubLibMock as any
			service.publishMessage(topic, data)
			expect(gcloudPubSubLibMock.publish).toHaveBeenCalled()
		})
	})
})
