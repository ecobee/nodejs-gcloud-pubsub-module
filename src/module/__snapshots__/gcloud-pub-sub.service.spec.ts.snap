// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GcloudPubSubService publishMessage Publishing a new message handles Buffer as data 1`] = `
{
  "data": [
    89,
    111,
    117,
    32,
    84,
    114,
    105,
    101,
    100,
    32,
    89,
    111,
    117,
    114,
    32,
    66,
    101,
    115,
    116,
    32,
    97,
    110,
    100,
    32,
    89,
    111,
    117,
    32,
    70,
    97,
    105,
    108,
    101,
    100,
    32,
    77,
    105,
    115,
    101,
    114,
    97,
    98,
    108,
    121,
    46,
    32,
    84,
    104,
    101,
    32,
    76,
    101,
    115,
    115,
    111,
    110,
    32,
    73,
    115,
    32,
    78,
    101,
    118,
    101,
    114,
    32,
    84,
    114,
    121,
  ],
  "type": "Buffer",
}
`;

exports[`GcloudPubSubService publishMessage Publishing a new message handles a SharedArrayBuffer as data 1`] = `
{
  "data": [
    0,
  ],
  "type": "Buffer",
}
`;

exports[`GcloudPubSubService publishMessage Publishing a new message handles a Uint8Array as data 1`] = `
{
  "data": [
    1,
    2,
    3,
  ],
  "type": "Buffer",
}
`;

exports[`GcloudPubSubService publishMessage Publishing a new message handles a string and binary encoding 1`] = `
{
  "data": [
    89,
    111,
    117,
    32,
    84,
    114,
    105,
    101,
    100,
    32,
    89,
    111,
    117,
    114,
    32,
    66,
    101,
    115,
    116,
    32,
    97,
    110,
    100,
    32,
    89,
    111,
    117,
    32,
    70,
    97,
    105,
    108,
    101,
    100,
    32,
    77,
    105,
    115,
    101,
    114,
    97,
    98,
    108,
    121,
    46,
    32,
    84,
    104,
    101,
    32,
    76,
    101,
    115,
    115,
    111,
    110,
    32,
    73,
    115,
    32,
    78,
    101,
    118,
    101,
    114,
    32,
    84,
    114,
    121,
  ],
  "type": "Buffer",
}
`;

exports[`GcloudPubSubService publishMessage Publishing a new message handles an ArrayBuffer as data 1`] = `
{
  "data": [
    0,
  ],
  "type": "Buffer",
}
`;

exports[`GcloudPubSubService publishMessage Publishing a new message handles an array of numbers as data 1`] = `
{
  "data": [
    10,
    20,
    30,
    40,
    50,
  ],
  "type": "Buffer",
}
`;

exports[`GcloudPubSubService publishMessage Publishing a new message handles string as data 1`] = `
{
  "data": [
    89,
    111,
    117,
    32,
    84,
    114,
    105,
    101,
    100,
    32,
    89,
    111,
    117,
    114,
    32,
    66,
    101,
    115,
    116,
    32,
    97,
    110,
    100,
    32,
    89,
    111,
    117,
    32,
    70,
    97,
    105,
    108,
    101,
    100,
    32,
    77,
    105,
    115,
    101,
    114,
    97,
    98,
    108,
    121,
    46,
    32,
    84,
    104,
    101,
    32,
    76,
    101,
    115,
    115,
    111,
    110,
    32,
    73,
    115,
    32,
    78,
    101,
    118,
    101,
    114,
    32,
    84,
    114,
    121,
  ],
  "type": "Buffer",
}
`;
